File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/DigitalCashHistoryFragment.java
package com.github.dedis.popstellar.ui.lao.digitalcash;

import android.os.Bundle;
import android.view.*;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.*;

import com.github.dedis.popstellar.R;
import com.github.dedis.popstellar.ui.lao.LaoActivity;
import com.github.dedis.popstellar.ui.lao.LaoViewModel;
import com.github.dedis.popstellar.utility.ActivityUtils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import io.reactivex.android.schedulers.AndroidSchedulers;

public class DigitalCashHistoryFragment extends Fragment {
  private static final Logger logger = LogManager.getLogger(DigitalCashHistoryFragment.class);

  private LaoViewModel laoViewModel;

  public static DigitalCashHistoryFragment newInstance() {
    return new DigitalCashHistoryFragment();
  }

  @Override
  public View onCreateView(
      LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    View view = inflater.inflate(R.layout.digital_cash_history_fragment, container, false);
    laoViewModel = LaoActivity.obtainViewModel(requireActivity());
    DigitalCashViewModel digitalCashViewModel =
        LaoActivity.obtainDigitalCashViewModel(getActivity(), laoViewModel.getLaoId());

    RecyclerView transactionList = view.findViewById(R.id.transaction_history_list);
    RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext());
    RecyclerView.ItemDecoration decoration =
        new DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL);

    HistoryListAdapter adapter = new HistoryListAdapter(digitalCashViewModel, requireActivity());

    transactionList.setLayoutManager(layoutManager);
    transactionList.addItemDecoration(decoration);
    transactionList.setAdapter(adapter);

    // Update dynamically the events in History
    laoViewModel.addDisposable(
        digitalCashViewModel
            .getTransactionsObservable()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                adapter::setList, error -> logger.debug("error with history update " + error)));

    handleBackNav();
    return view;
  }

  @Override
  public void onResume() {
    super.onResume();
    laoViewModel.setPageTitle(R.string.digital_cash_history);
    laoViewModel.setIsTab(false);
  }

  private void handleBackNav() {
    LaoActivity.addBackNavigationCallback(
        requireActivity(),
        getViewLifecycleOwner(),
        ActivityUtils.buildBackButtonCallback(
            logger, "last fragment", ((LaoActivity) requireActivity())::resetLastFragment));
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/DigitalCashHistoryFragment.kt
package com.github.dedis.popstellar.ui.lao.digitalcash

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.github.dedis.popstellar.R
import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObject
import com.github.dedis.popstellar.ui.lao.LaoActivity
import com.github.dedis.popstellar.ui.lao.LaoActivity.Companion.addBackNavigationCallback
import com.github.dedis.popstellar.ui.lao.LaoActivity.Companion.obtainDigitalCashViewModel
import com.github.dedis.popstellar.ui.lao.LaoActivity.Companion.obtainViewModel
import com.github.dedis.popstellar.ui.lao.LaoViewModel
import com.github.dedis.popstellar.utility.ActivityUtils.buildBackButtonCallback
import io.reactivex.android.schedulers.AndroidSchedulers
import timber.log.Timber

class DigitalCashHistoryFragment : Fragment() {

  private lateinit var laoViewModel: LaoViewModel

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View? {
    // Inflate the layout for this fragment
    val view = inflater.inflate(R.layout.digital_cash_history_fragment, container, false)

    laoViewModel = obtainViewModel(requireActivity())
    val digitalCashViewModel = obtainDigitalCashViewModel(requireActivity(), laoViewModel.laoId!!)

    val transactionList = view.findViewById<RecyclerView>(R.id.transaction_history_list)

    val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(context)
    val decoration: ItemDecoration =
        DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL)
    val adapter = HistoryListAdapter(digitalCashViewModel, requireActivity())

    transactionList.layoutManager = layoutManager
    transactionList.addItemDecoration(decoration)
    transactionList.adapter = adapter

    // Update dynamically the events in History
    laoViewModel.addDisposable(
        digitalCashViewModel.transactionsObservable
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { transactions: List<TransactionObject> -> adapter.setList(transactions) },
                { error: Throwable -> Timber.tag(TAG).e(error, "error with history update") }))

    handleBackNav()

    return view
  }

  override fun onResume() {
    super.onResume()
    laoViewModel.setPageTitle(R.string.digital_cash_history)
    laoViewModel.setIsTab(false)
  }

  private fun handleBackNav() {
    addBackNavigationCallback(
        requireActivity(),
        viewLifecycleOwner,
        buildBackButtonCallback(TAG, "last fragment") {
          (requireActivity() as LaoActivity).resetLastFragment()
        })
  }

  companion object {
    private val TAG = DigitalCashHistoryFragment::class.java.simpleName

    fun newInstance(): DigitalCashHistoryFragment {
      return DigitalCashHistoryFragment()
    }
  }
}
